flink-metrics/flink-metrics-jmx/src/main/java/org/apache/flink/metrics/jmx/JMXReporter.java

```diff
@@ -541,5 +552,40 @@ public void stop() throws IOException {
 				}
 			}
 		}
+
+		/*
+		 * Better to use the internal API than re-invent the wheel.
+		 */
+		@SuppressWarnings("restriction")
+		private class JmxRegistry extends sun.rmi.registry.RegistryImpl {
+			private final String lookupName;
+
+			JmxRegistry(final int port, final String lookupName) throws RemoteException {
+				super(port);
+				this.lookupName = lookupName;
+			}
+
+			@Override
+			public Remote lookup(String s) throws RemoteException, NotBoundException {
+				return lookupName.equals(s) ? remoteServerStub : null;
```

Comment:
The `remoteServerStub` would be null between starting the registry and starting the server. Probably not a big issue in practice, but looks like the separation between `startRmiRegistry()` and `startJmxService()` does not work well here.
Looks almost like the `startJmxService()` should happen before the `startRmiRegistry()` and then pass the stub to the registry. But I don't know the internals of JMX well enough to judge whether this makes sense.

Reply:
Yes, good point, this could probably be improved a bit.

Reply:
Yes, good point, this could probably be improved a bit.